<launch>
    
    <!-- ROSSerial with Teensy -->
    <arg name="teensy_path" default="/dev/teensy" />
    <arg name="teensy_baudrate" default="115200" />
    <node
        name="teensy" pkg="rosserial_python" type="serial_node.py" respawn="true" output="screen">
        <param name="port" value="$(arg teensy_path)" />
        <param name="baud" value="$(arg teensy_baudrate)" />
    </node>

    <!-- LIDAR -->
    <arg name="lidar_path" default="/dev/Hokuyo" />
    <arg name="lidar_frame" default="base_lidar" />
    <node
        name="hokuyo" pkg="urg_node" type="urg_node" respawn="true" output="screen">
        <param name="serial_port" value="$(arg lidar_path)" />
        <param name="frame_id" value="$(arg lidar_frame)" />
    </node>
    
    <!-- IMU -->
    <arg name="imu_path" default="/dev/VMU931" />
    <arg name="imu_baudrate" default="9600" />
    <node
        name="vmu931" pkg="vmu931" type="vmu931_node" respawn="true" output="screen">
        <param name="serial_port" value="$(arg imu_path)" />
        <param name="serial_baudrate" value="$(arg imu_baudrate)" />
    </node>

    <!-- TF Publisher -->
    <node name="static_tf_broadcaster" pkg="tf_broadcaster" type="static_tf_broadcaster" respawn="true" />
    <arg name="odom_tf" default="false" />
    <arg name="odom_covariance" default="0.1" />
    <node
        name="odom_publisher" pkg="tf_broadcaster" type="odom_publisher" respawn="true">
        <param name="odom_covariance" value="$(arg odom_covariance)" />
        <param name="odom_tf" value="$(arg odom_tf)" />
    </node>
    
    <!-- EKF -->
    <arg name="use_ekf" default="true" />
    <arg name="ekf_tf" default="false" />
    <group if="$(arg use_ekf)">
        <include file="$(find robot_pose_ekf)/robot_pose_ekf.launch">
            <arg name="base_footprint_frame" value="base_link" if="$(arg ekf_tf)" />
            <arg name="odom_used" value="true" />
            <arg name="imu_used" value="true" />
            <arg name="vo_used" value="false" />
            <arg name="remap_frame" value="odom" />
            <arg name="remap_imu" value="imu" />
        </include>
    </group>
    
    <!-- Laser Scan Matcher -->
    <arg name="use_scan_matcher" default="true" />
    <arg name="scan_matcher_tf" default="true" />
    <group if="$(arg use_scan_matcher)">
        <node
            name="scan_matcher" pkg="laser_scan_matcher" type="laser_scan_matcher_node" output="screen">
            <param name="fixed_frame" value="map" />
            <param name="use_imu" value="true" />
            <param name="use_odom" value="true" />
            <param name="use_vel" value="false" />
            <param name="publish_tf" value="false" unless="$(arg scan_matcher_tf)" />
            <param name="publish_pose" value="false" />
            <param name="publish_pose_with_covariance_stamped" value="true" />
            <remap from="odom" to="robot_pose_ekf/odom_combined" if="$(arg use_ekf)" />
            <remap from="imu/data" to="imu" />
        </node>
    </group>
    
    <!-- Joy input -->
    <arg name="use_joy" default="true" />
    <node name="joy_listener" pkg="send_cmd" type="joy_listener" if="$(arg use_joy)" />
    
    <!-- Combine map and odom frames to view both in rviz -->
    <arg name="combine_frames" default="false" />
    <node if="$(arg combine_frames)" name="tf_combine_frames" pkg="tf_broadcaster" type="tf_combine_frames" output="screen" />
    
    <!-- Rviz -->
    <arg name="use_rviz" default="false" />
    <group if="$(arg use_rviz)">
        <param name="robot_description" textfile="$(find ffast_urdf)/robots/ffast_urdf.urdf" />
        <node name="rviz" pkg="rviz" type="rviz" />
    </group>
    
</launch>
